pkg/healthcheck/healthcheck.go

```diff
@@ -2191,6 +2209,90 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
 	return nil
 }
 
+// Check if there's a pod with the opaque-ports annotation defined but a
+// service selecting the aforementioned pod doesn't define it
+func (hc *HealthChecker) checkMisconfiguredOpaquePortAnnotations(ctx context.Context, namespace string) error {
+	services, err := hc.GetServices(ctx)
+	if err != nil {
+		return err
+	}
+
+	pods, err := hc.kubeAPI.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	var errStrings []string
+
+	for _, service := range services {
+		if mismatch := misconfiguredOpaquePortAnnotationsInService(service, pods); mismatch {
+			errStrings = append(
+				errStrings,
+				fmt.Sprintf("%s annotation is not properly configured in service %s",
+					k8s.ProxyOpaquePortsAnnotation,
+					service.Name),
+			)
+		}
+	}
+
+	if len(errStrings) >= 1 {
+		return fmt.Errorf(strings.Join(errStrings, "\n"))
+	}
+
+	return nil
+}
+
+func misconfiguredOpaquePortAnnotationsInService(service corev1.Service, pods *corev1.PodList) bool {
+	for _, pod := range pods.Items {
+		// The endpoints for headless services are handled differently so don't check this type of services
+		if service.Spec.ClusterIP != "None" && serviceMatchesPod(service, pod) {
+			return missconfiguredOpaqueAnnotation(service.Annotations, pod.Annotations)
+		}
+	}
+	return false
+}
+
+func missconfiguredOpaqueAnnotation(svcAnnotations map[string]string, podAnnotations map[string]string) bool {
+	// If the pod has the annotation, check that the service has it as well
+	if podAnnotation, found := podAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+		if svcAnnotation, found := svcAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+			if svcAnnotation != podAnnotation {
+				return true
+			}
+		} else {
+			return true
+		}
+	}
+
+	// If the service has the annotation, check that the pod has it as well
+	if svcAnnotation, found := svcAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+		if podAnnotation, found := podAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+			if svcAnnotation != podAnnotation {
+				return true
+			}
+		} else {
+			return true
+		}
+	}
+
+	return false
+}
+
+func serviceMatchesPod(service corev1.Service, pod corev1.Pod) bool {
+	podSelector := service.Spec.Selector
+	for k, sv := range podSelector {
+		if pv, ok := pod.Labels[k]; ok {
+			if sv != pv {
+				return false
+			}
+		} else {
+			return false
+		}
```

Comment:
hm, so here it seems that we return after checking just one label? the way I'm reading this: for the labels in the service's selector, if the pod includes that label in its map and if it doesn't match the service we return false. If the pod does not have that label we still return false. should we instead skip this loop iteration? seems that we could return early from this function and miss out on some processing?

Reply:
Not sure I follow you, sorry. We need to go through all the labels in the service selector to make sure __all__ of them match.

So, if a service has defined a selector like

```
  selector:
    app: app
    tier: backend
```

we need to iterate through all the labels; if we don't find one of them in the pod or find it but the values don't match, we can return early because we know that the service will not match (selectors are combined using AND).

So, if we remove the loop, we wouldn't be checking the full selector, would we?

Reply:
Ahhhhh, I see. That makes sense!

Reply:
Ahhhhh, I see. That makes sense!